File: bindings/swift/Sources/Causalontology/Jcs.swift
// Jcs.swift
//
// RFC 8785 (JSON Canonicalization Scheme) serialization.
//
// A faithful port of the Python binding's serializer for the value ranges
// Causalontology uses (integers, integer-valued floats, and short decimals);
// full ECMAScript exponent formatting for extreme magnitudes is pinned at
// the 1.0.0 conformance freeze, exactly as in the Python binding.

import Foundation

/// The canonical serialization of a JsonValue as a String.
public func jcsString(_ value: JsonValue) throws -> String {
    switch value {
    case .null:
        return "null"
    case let .bool(flag):
        return flag ? "true" : "false"
    case let .int(number):
        return String(number)
    case let .double(number):
        return try jcsNumber(number)
    case let .string(text):
        return jcsQuotedString(text)
    case let .array(items):
        var parts: [String] = []
        for item in items {
            parts.append(try jcsString(item))
        }
        return "[" + parts.joined(separator: ",") + "]"
    case let .object(members):
        // RFC 8785: member names sorted by their UTF-16 code units.
        let sortedKeys = members.keys.sorted(by: utf16LessThan)
        var parts: [String] = []
        for key in sortedKeys {
            let serializedValue = try jcsString(members[key]!)
            parts.append(jcsQuotedString(key) + ":" + serializedValue)
        }
        return "{" + parts.joined(separator: ",") + "}"
    }
}

/// The canonical serialization of a JsonValue as UTF-8 bytes.
public func jcsData(_ value: JsonValue) throws -> Data {
    return Data(try jcsString(value).utf8)
}

/// True when a sorts strictly before b by UTF-16 code units (RFC 8785).
func utf16LessThan(_ a: String, _ b: String) -> Bool {
    let aUnits = Array(a.utf16)
    let bUnits = Array(b.utf16)
    let shared = min(aUnits.count, bUnits.count)
    var i = 0
    while i < shared {
        if aUnits[i] != bUnits[i] {
            return aUnits[i] < bUnits[i]
        }
        i += 1
    }
    return aUnits.count < bUnits.count
}

/// The RFC 8785 string form: minimal escaping only.
func jcsQuotedString(_ text: String) -> String {
    var out = "\""
    for scalar in text.unicodeScalars {
        switch scalar.value {
        case 0x22:
            // The double quote.
            out += "\\\""
        case 0x5C:
            // The backslash.
            out += "\\\\"
        case 0x08:
            out += "\\b"
        case 0x09:
            out += "\\t"
        case 0x0A:
            out += "\\n"
        case 0x0C:
            out += "\\f"
        case 0x0D:
            out += "\\r"
        default:
            if scalar.value < 0x20 {
                // Other control characters as lowercase \u00xx.
                out += String(format: "\\u%04x", scalar.value)
            } else {
                out.append(Character(scalar))
            }
        }
    }
    out += "\""
    return out
}

/// The RFC 8785 number form for a floating value, mirroring the Python
/// binding's _jcs_number: integer-valued floats below 1e21 print with no
/// decimal point and no exponent; other values use the shortest round-trip
/// decimal with a normalized ECMAScript-style exponent.
func jcsNumber(_ number: Double) throws -> String {
    guard number.isFinite else {
        throw CausalontologyError("NaN and Infinity are not permitted (RFC 8785)")
    }
    // Covers both 0.0 and -0.0, which canonicalize to "0".
    if number == 0 {
        return "0"
    }
    if number.truncatingRemainder(dividingBy: 1) == 0 && abs(number) < 1e21 {
        // Integer-valued: print the integer digits. Every integer-valued
        // Double with magnitude below 2^63 is exactly representable here;
        // the (unused-in-practice) band up to 1e21 falls through below.
        if let integer = Int64(exactly: number) {
            return String(integer)
        }
    }
    // Swift's Double description is the shortest decimal that round-trips,
    // matching ECMAScript digits for this range.
    var text = "\(number)"
    if let eIndex = text.firstIndex(of: "e") {
        // Normalize the exponent: 1e-07 -> 1e-7; positive keeps '+'.
        let mantissa = String(text[text.startIndex..<eIndex])
        var exponent = String(text[text.index(after: eIndex)...])
        var sign = "+"
        if exponent.hasPrefix("-") {
            sign = "-"
            exponent.removeFirst()
        } else if exponent.hasPrefix("+") {
            exponent.removeFirst()
        }
        while exponent.count > 1 && exponent.hasPrefix("0") {
            exponent.removeFirst()
        }
        text = mantissa + "e" + sign + exponent
    }
    return text
}

File: bindings/r/R/jcs.R
# causalontology-r -- jcs.R
#
# RFC 8785 (JSON Canonicalization Scheme) serialization, ported from the
# reference bindings/python/causalontology/canonical.py (_jcs and friends).
#
# Number policy (matching the reference for the value ranges the standard
# uses -- see the note in canonical.py):
#   integer-shaped (attr co_shape == "int")        -> sprintf("%.0f")
#   float-shaped but integral and |x| < 1e21       -> sprintf("%.0f")
#   otherwise                                      -> shortest round-trip
#     decimal, with C-style exponents ("1e-07") normalized to the
#     ECMAScript form ("1e-7") exactly as the reference does.
# Untagged doubles (values built by R code rather than parsed from JSON)
# fall through to the same by-value rules, which reproduce the reference
# output for every value in the suite (0 -> "0", 0.8 -> "0.8").

# Minimal string escaping per RFC 8785: the two mandatory escapes, the
# short names for the common controls, \u00xx for the rest below 0x20.
co_jcs_string <- function(s) {
  chars <- strsplit(enc2utf8(s), "", fixed = FALSE)[[1]]
  out <- character(length(chars))
  for (i in seq_along(chars)) {
    ch <- chars[[i]]
    if (identical(ch, "\"")) {
      out[[i]] <- "\\\""
    } else if (identical(ch, "\\")) {
      out[[i]] <- "\\\\"
    } else if (identical(ch, "\b")) {
      out[[i]] <- "\\b"
    } else if (identical(ch, "\t")) {
      out[[i]] <- "\\t"
    } else if (identical(ch, "\n")) {
      out[[i]] <- "\\n"
    } else if (identical(ch, "\f")) {
      out[[i]] <- "\\f"
    } else if (identical(ch, "\r")) {
      out[[i]] <- "\\r"
    } else {
      cp <- utf8ToInt(ch)
      if (cp < 0x20) {
        out[[i]] <- sprintf("\\u%04x", cp)
      } else {
        out[[i]] <- ch
      }
    }
  }
  paste0("\"", paste(out, collapse = ""), "\"")
}

# The shortest decimal string that round-trips to exactly x (the analogue
# of Python's repr / ECMAScript's Number::toString for our value range).
co_shortest_double <- function(x) {
  for (digits in 1:17) {
    s <- sprintf("%.*g", digits, x)
    if (isTRUE(as.numeric(s) == x)) return(s)
  }
  sprintf("%.17g", x)
}

co_jcs_number <- function(n) {
  # booleans are handled by co_jcs before this point
  shape <- attr(n, "co_shape")
  v <- as.numeric(n)[[1]]
  if (!is.finite(v)) stop("NaN and Infinity are not permitted (RFC 8785)")
  if (identical(shape, "int")) {
    return(sprintf("%.0f", v))          # exact for |v| <= 2^53
  }
  if (isTRUE(v == 0)) return("0")
  if (isTRUE(v == floor(v)) && abs(v) < 1e21) {
    return(sprintf("%.0f", v))          # 1.0 -> "1", 6.000 -> "6"
  }
  s <- co_shortest_double(v)            # 0.7 -> "0.7"
  if (grepl("e", s, fixed = TRUE)) {    # normalize exponent: 1e-07 -> 1e-7
    parts <- strsplit(s, "e", fixed = TRUE)[[1]]
    mant <- parts[[1]]
    ex <- parts[[2]]
    sign <- if (identical(substring(ex, 1L, 1L), "-")) "-" else "+"
    digits <- sub("^[+-]?0*", "", ex)
    if (!nzchar(digits)) digits <- "0"
    s <- paste0(mant, "e", sign, digits)
  }
  s
}

# Serialize one value in RFC 8785 canonical form.
co_jcs <- function(value) {
  if (is.null(value) || co_is_null(value)) return("null")
  if (co_is_obj(value)) {
    keys <- names(value)
    if (is.null(keys)) keys <- character(0)
    keys <- co_sort_codepoint(keys)     # sorted keys, code-point order
    parts <- character(length(keys))
    for (i in seq_along(keys)) {
      k <- keys[[i]]
      parts[[i]] <- paste0(co_jcs_string(k), ":", co_jcs(value[[k]]))
    }
    return(paste0("{", paste(parts, collapse = ","), "}"))
  }
  if (co_is_arr(value)) {
    parts <- character(length(value))
    for (i in seq_along(value)) parts[[i]] <- co_jcs(value[[i]])
    return(paste0("[", paste(parts, collapse = ","), "]"))
  }
  if (is.logical(value)) return(if (isTRUE(value[[1]])) "true" else "false")
  if (is.numeric(value)) return(co_jcs_number(value))
  if (is.character(value)) return(co_jcs_string(value[[1]]))
  stop("cannot canonicalize value of class ", paste(class(value), collapse = "/"))
}
